### ISSUE TITLE
Next.js v12 compatibility

### ISSUE THREAD

**Issue by toomuchdesign** [next-page-tester/next-page-tester#281]
## 🐛 Bug Report

As [this PR shows](https://github.com/toomuchdesign/next-page-tester/pull/280), `next-page-tester` has issues with **Next.js v12**. At a first sight I found:

- some types issues
- an unexpected `segmentation fault` error while running tests

I can't predict when I'll be able to work on it. As usual contributions are very welcome 🙌 

## Reproduction

https://github.com/toomuchdesign/next-page-tester/pull/280

**Comment by movahedan** [comment]
I don't know it is the exact issue you mentioned here or not,
I get this `Segmentation fault (core dumped)` when I'm trying to test my pages after upgrading to next@12.0.1

This is my test:
```
        it('page renders and matches snapshot', async () => {
		const { serverRenderToString, render } = await getPage({
			route: '/_error',
		});

		// check correctness of SSR result
		const { html } = serverRenderToString();
		expect(html).toMatchSnapshot();

		// check hydrated app
		const { nextRoot } = render();
		expect(nextRoot.outerHTML).toMatchSnapshot();
	});
```

**Comment by toomuchdesign** [comment]
Hi @movahedan,
thanks for your feedback. Yep I also got a similar error. I suspect the cause of it might be some Next.js internal refactor.

**Comment by movahedan** [comment]
@toomuchdesign Yes, it might be related. Storybook seems to have conflicts with Next.js 12 too, they are not updating themselves into webpack 5. 
I'm too amateur in testing and jest, so I can't raise a PR for this, sorry for that. I hope you find some time to fix it.

**Comment by GheorgheP** [comment]
I have the same issue. Tested with a clean app created using `yarn create next-app`.

**Comment by calbrecht** [comment]
Maybe it has to do with https://github.com/nodejs/node/issues/40602. Debugging into [await loadConfig](https://github.com/toomuchdesign/next-page-tester/blob/c34416c110bd1f02812b2685675e585f9111591f/src/nextConfig.ts#L37) where it tries to load the `next.config.js` file by calling [await import](https://github.com/vercel/next.js/blob/4329b7f2b5b76136e7e258b5b11c38b8fc68e7f7/packages/next/server/config.ts#L555) is leading to the segfault.

**Comment by calbrecht** [comment]
https://github.com/vercel/next.js/pull/29935 seem to have caused it and the followup https://github.com/vercel/next.js/pull/29971 mentions the segfault on node 14 for  other platforms than linux. However, the platform i am experiencing this behaviour is a NixOS with node 16.13.0.

**Comment by braccolucianoj** [comment]
Hello, I ran into the same issue and I was able to fix using jest moduleNameMapper. The issue it's related to the default function exported from this file `next/dist/server/config`. Also since `next-page-tester` doesn't use the Function attributes exported form next config, this import can be easily "mocked" by just exporting the serialized representation of a result of a dev run of the configuration.

So in jest.config file add an entry on the moduleNameMapper `'next/dist/server/config': 'path/to/mocked/next/config/loader.js` and on this file add the object as `export default () => (configObject)`

That worked for me! Hope this helps

**Comment by jasonwilliams** [comment]
I haven't had segfault issues instead my issue has been around the next/dist/server/config

```
  ● Test suite failed to run

    Cannot find module 'next/dist/server/config' from '../node_modules/next-page-tester/dist/nextConfig.js'

    Require stack:
      /user/project/ui/node_modules/next-page-tester/dist/nextConfig.js
      /user/project/ui/node_modules/next-page-tester/dist/utils.js
      /user/project/ui/node_modules/next-page-tester/dist/getPageObject/makeRouteInfo.js
      /user/project/ui/node_modules/next-page-tester/dist/getPageObject/index.js
      /user/project/ui/node_modules/next-page-tester/dist/page/getPageInfo.js
      /user/project/ui/node_modules/next-page-tester/dist/page/index.js
      /user/project/ui/node_modules/next-page-tester/dist/getPage.js
      /user/project/ui/node_modules/next-page-tester/dist/pure.js
      /user/project/ui/node_modules/next-page-tester/dist/index.js
      tests/jest.setup.ts

      at Resolver.resolveModule (../node_modules/jest-resolve/build/resolver.js:311:11)
      at Object.<anonymous> (../node_modules/next-page-tester/dist/nextConfig.js:8:34)
```

**Comment by dbehmoaras** [comment]
Maybe this is also an issue with Next 12. Next-page-tester's render method does not render anything and ignores assertions. The test just simply stops running after invoking the render method.

**Comment by jasonwilliams** [comment]
@toomuchdesign is there anything we can do to help?

**Comment by toomuchdesign** [comment]
Hi @jasonwilliams, unfortunately in the near future I won't be able to spend time on `next-page-tester`. I've opened #290 to discuss a way to proceed further.

**Comment by jasonwilliams** [comment]
@toomuchdesign do you know why v12 doesn’t work? I think we can sort out a fork and fix it but in this thread I’m keen to know why it stopped working in this release.

**Comment by toomuchdesign** [comment]
Unfortunately I haven't had the opportunity to dig into Next 12 changes.

**Comment by jasonwilliams** [comment]
I hope to take a look at this again soon, but won’t be for a while. If anyone has any useful info feel free to add here. From looking at the upstream issues it seems node hasn’t fixed the segfaults

**Comment by jasonwilliams** [comment]
> I don't know if it is the exact issue you mentioned here or not, I get this `Segmentation fault (core dumped)` when I'm trying to test my pages after upgrading to next@12.0.2
> 
> This is my test:
> 
> ```
>         it('page renders and matches snapshot', async () => {
> 		const { serverRenderToString, render } = await getPage({
> 			route: '/_error',
> 		});
> 
> 		// check correctness of SSR result
> 		const { html } = serverRenderToString();
> 		expect(html).toMatchSnapshot();
> 
> 		// check hydrated app
> 		const { nextRoot } = render();
> 		expect(nextRoot.outerHTML).toMatchSnapshot();
> 	});
> ```
> 
> This information could be useful:
> 
> ```
> "next": "^12.0.2",
> "react": "^17.0.2",
> "react-dom": "^17.0.2",
> "jest": "^27.1.0",
> "babel-jest": "^27.1.0",
> ```

Crazy idea but does moving to viteTest fix this?
It seems like this segmentation fault is still with jest and not fixed yet, but im unsure if ViteTest is affected

**Comment by jasonwilliams** [comment]
There's a new PR open for this. The type errors have been sorted for now https://github.com/next-page-tester/next-page-tester/pull/294 Will need help deciding what to do about Jest though.

**Comment by jasonwilliams** [comment]
Just to update everyone on what's going on here..

The segfaults look to be coming from a mix of require and import, plus the way Jest operates.
Considering Jest is not [getting supported](https://news.ycombinator.com/item?id=30168241) much anymore we will need a long term plan to move away from Jest completely. This means we have 2 PRs..

## How can I help?

### Short Term
The short term fix is to resolve the Type Issues and segfaults in Jest tests. This is https://github.com/next-page-tester/next-page-tester/pull/294

It looks like we've fixed the main issue by swapping to import in some places. The case here is to try and test this PR out and give feedback. If we can get something out without breaking other test suites that would be good to know.

### Long Term
The long term fix is to migrate to Vitest. The PR is here https://github.com/next-page-tester/next-page-tester/pull/295
You can help by checking that out and testing.

### Do you use something other than Jest?

Your help is needed in https://github.com/next-page-tester/next-page-tester/pull/294

**Comment by IGassmann** [comment]
@jasonwilliams it's too early for concluding that Jest won't be supported anymore. Considering how popular is the package, it's likely that someone else or another organization might start leading its contributions.

**Comment by jasonwilliams** [comment]
@IGassmann you're right, this isn't about dropping support for Jest per-say. But providing us a path to move forward and not being blocked by what's happening with Jest.

Internally this could change its test suite and still work with other test runners people use. It's true Vitest is in development right now, but the support for native ESM Modules is a direction we want to head in, so building against that works quite nicely right now.

**Comment by toomuchdesign** [comment]
Hi @jasonwilliams. have you had the opportunity to get what is exactly causing the esm/commonjs hell? Was setting the `__NEXT_TEST_MODE` flag enough?

I tried to understand more about the nature of the issue and found [this article](https://redfin.engineering/node-modules-at-war-why-commonjs-and-es-modules-cant-get-along-9617135eeca1) very well written.
